searchcore/src/vespa/searchcore/proton/matching/matcher.cpp

```diff
@@ -311,16 +311,19 @@ Matcher::match(const SearchRequest &request, vespalib::ThreadBundle &threadBundl
     total_matching_time.stop();
     my_stats.queryCollateralTime(total_matching_time.elapsed().sec() - my_stats.queryLatencyAvg());
     {
-        fastos::TimeStamp softLimit = uint64_t((1.0 - _rankSetup->getSoftTimeoutTailCost()) * request.getTimeout());
         fastos::TimeStamp duration = request.getTimeUsed();
         std::lock_guard<std::mutex> guard(_statsLock);
         _stats.add(my_stats);
         if (my_stats.softDoomed()) {
             double old = _stats.softDoomFactor();
+            fastos::TimeStamp left = my_stats.timeLeftAtDoom();
+            fastos::TimeStamp realTimeout = (left >= 0) ? request.getTimeout() : (request.getTimeout() - left);
+            fastos::TimeStamp softLimit = uint64_t((1.0 - _rankSetup->getSoftTimeoutTailCost()) * realTimeout);
```

Comment:
setting the soft limit higher than the hard limit will break basic assumptions. I think a better solution is to simply subtract 'max(doomOvertime)' from 'duration' to ignore the fact that we did not bail out when we wanted to, in order to focus more on the time we need to clean up after ourself in order to send a partial result. We could also consider only ignoring overtime exceeding some small number...

Reply:
Done.